src/context/common.ml

```diff
@@ -611,20 +611,42 @@ let find_file ctx f =
 	with Exit ->
 		raise Not_found
 	| Not_found ->
+		let f_dir = Filename.dirname f
+		and platform_ext = "." ^ (platform_name_macro ctx)
+		and is_core_api = defined ctx Define.CoreApi in
 		let rec loop had_empty = function
 			| [] when had_empty -> raise Not_found
 			| [] -> loop true [""]
 			| p :: l ->
 				let file = p ^ f in
-				if Sys.file_exists file then begin
-					(try
-						let ext = String.rindex file '.' in
-						let file_pf = String.sub file 0 (ext + 1) ^ platform_name_macro ctx ^ String.sub file ext (String.length file - ext) in
-						if not (defined ctx Define.CoreApi) && Sys.file_exists file_pf then file_pf else file
-					with Not_found ->
-						file)
-				end else
-					loop (had_empty || p = "") l
+				let dir = Filename.dirname file in
+				let found = ref "" in
+				(try
+					Array.iter
+						(fun file_name ->
+							let current_f = if f_dir = "." then file_name else Filename.concat f_dir file_name in
+							let is_cached = Hashtbl.mem ctx.file_lookup_cache current_f in
+							if
+								is_core_api
+								|| (
+									platform_ext = Filename.extension (Filename.remove_extension file_name)
+									|| not is_cached
+								)
+							then begin
+								let full_path = Filename.concat dir file_name in
+								if is_cached then
+									Hashtbl.remove ctx.file_lookup_cache current_f;
+								Hashtbl.add ctx.file_lookup_cache current_f (Some full_path);
```

Comment:
Is this remove + add in any way better than using replace?

Reply:
It's a premature optimization :)
Hashtbl doc says `replace` = `remove + add`, and since we already have `is_cached` we can avoid an item lookup for `remove`.